### PR TITLE
refactor(neard): decouple into individual crates

### LINKED ISSUE #4233: Decouple `nearcore` into a lib-crate
As part of the bigger initiative #1881, we want to break `neard` crate into two:

* `neard` binary
* `nearcore` library

We use `neard` as a library in Indexer Framework, and it might also improve testability and maintainability. This will also lay a path towards better dependencies handling.

### PR DESCRIPTION
**Pull request by miraclx**
Dissociates `neard` into its binary and library constituents, with the binary depending on the library. Fixes #4233.

The intent is to have the binary named `neard` and the library named `nearcore`. However, because the workspace is itself `nearcore`, we can either move the library into the root or use a different name for either the workspace or the library itself.

### CODE REVIEW

File: neard/bin/Cargo.toml
Comment (bowenwang1996) [inline neard/bin/Cargo.toml:2]: Let's keep using `neard` as the name of the binary. Otherwise a bunch of docs and code that rely on this will break

Comment (miraclx) [reply neard/bin/Cargo.toml:2]: Understood. And yeah, that's the plan.
Although within the workspace, a lot equally depends on `neard` as the lib name. With the _first pass_, it's like this for now while we make sure we have a proper structure for where we want the lib to go.

The options available are:
- Have the lib named `nearcore`, for which we have two options:
  - Move the lib into the root crate.
  - Rename the workspace.
- Rename the lib. I mentioned something like `libnear` but @frol rightly pointed out that seems more appropriate for a C interop we might not want to block by taking that name.

WDYT?

**Comment by matklad** [comment]
>However, because the workspace is itself nearcore

I feel that that's unfortunate. The root of our workspace is a crate, but this crate only contains integration tests, it isn't a binary or a library. That's surprising (most big projects are set-up differently) and obscures the actual dependencies between the things. 

The typical setup is to make the root of the workspace a "virtual manifest", without the `[package]` section. See examples: [rust-analyzer](https://github.com/rust-analyzer/rust-analyzer/blob/master/Cargo.toml), [rustc](https://github.com/rust-lang/rust/blob/master/Cargo.toml), [substrate](https://github.com/paritytech/substrate/blob/master/Cargo.toml), [firecracker](https://github.com/firecracker-microvm/firecracker/blob/main/Cargo.toml). That way, you don't have "primary" crate, which often makes sense for large projects, as they have several entry points. `cargo test` in the workspace by default test everything. And you get somewhat more sane `--feature` flag behavior (with the second version of feature resolver, we should be able to even get rid of `foo = [bar/foo, baz/foo]` feature propagations).

Large project with non-virtual manifest typically have a top-level crate of substance, with `src` top-level directory which contains stuff, like [cargo](https://github.com/rust-lang/cargo).

So, if we want to introduce a genuine `nearcore` library, I would suggest to get rid of tests-only top-level package first. Between "let's make top-level `nearcore` a real library" and "let's have a virtual manifest", I think "virtual manifest" has more benefits, as we do have several entry points (neard, params estimator, tools like genesis-populate). 

I understand that this derails immediate work of extracing a library out of neard, but "fix workspace, *then* split neard" seems to be less work overall.

**Comment by matklad** [comment]
Here's a my suggested plan for virtual manifest:

* check which tooling depends on the project layout and might break. Quick look at `pipeline.yml` shows that it should work with virtual manifest as is, but we might have some other automation.a
* move tests from `/test` folder somewhere -- either to some of the existing packages (neard perhaps?) or into a new `integration-tests` package
* re-read the features chapter of the cargo book, especially this section: https://doc.rust-lang.org/cargo/reference/features.html#resolver-version-2-command-line-flags
* check that features in top-level toml will work when we switch to virtual manifest (with resolevr = "2" that should be the case)
* add `[workspace] resolver = "2"` to the top-level toml
* remove `[package]`, `[dependencies]`, `[features]` section from top-level toml (`[profile]` should stay)
* check that everything works as expected. 

Is this reasonable? Let's do a checklist to make sure we decide one way or another:

Do you agree that we should move to virtual manifest setup *before* introducing proper `nearcore` library? 

- [ ] @bowenwang1996 
- [ ] @frol 

(tagging a couple of folks who I know have opinions about code organization)

**Comment by frol** [comment]
I totally agree with the proposals from @matklad.

I don't expect any breakage from dropping `[package]` from the workspace Cargo.toml, and I even consider it is a great thing to do since it was quite confusing when you would call `cargo test` or `cargo build` which did nothing useful.

I suggest we move the current `tests/` into `nearcore` (lib crate) if possible, otherwise fall back to `neard` (bin).

**Review by frol — APPROVED** [review]
We are on the right track! Once CI is green, we can merge it

**Review by chefsale — APPROVED** [review]
LGTM

**Review by ailisp — APPROVED** [review]
Looks good!

File: .buildkite/pipeline.yml
Comment (matklad) [inline .buildkite/pipeline.yml:127]: Thanks for clarifying what this does!

File: chain/client/Cargo.toml
Comment (matklad) [inline chain/client/Cargo.toml:9]: :+1: it's good idea to pin pre-releases which are used in production.

Unrelated to this PR, but I've noticed that we need to update this in a lot of Cargo.tomls. It seems that some deps shouldn't be there? For example, the one in chain/chunks seems to be removable.

File: neard/Cargo.toml
Comment (matklad) [inline neard/Cargo.toml:14]: Unrelated to the PR, but there's a chance we no-longer need this work-around. If that's the case, we can remove both bins.

File: neard/Cargo.toml
Comment (matklad) [inline neard/Cargo.toml:58]: We probably want all this features to be 

```
feature_name = ["nearcore/feature_name"]
```

This can be landed in a follow up PR, but it definitely has to happen as soon as possible, otherwise feature lists will diverge.

**Review by matklad — APPROVED** [review]

File: neard/Cargo.toml
Comment (frol) [reply neard/Cargo.toml:14]: Good point, but I want this PR to keep things identical to the previous behavior wherever possible, so we don't spend time on unrelated discussions and don't revert the whole refactoring if we happen to decide to revert this change for some reason. Indeed, I plan to submit a PR dropping the old `near` binary.

**Review by khorolets — APPROVED** [review]
LGTM!

**Review by bowenwang1996 — APPROVED** [review]

File: Cargo.toml
Comment (olonho) [inline Cargo.toml:130]: Those features are needed, why they are removed?

Comment (matklad) [reply Cargo.toml:130]: Good catch, I didn't notice that! I think it'd be good indeed to move this over to `neard/Cargo.toml` instead of just deleting. 

That being said, I wouldn't block on that, as the following now works without feature forwarding:

```
cargo build -p neard --features node-runtime/wasmer1_default
```

Comment (frol) [reply Cargo.toml:130]: @olonho Great catch!

It seems it should be in both `nearcore` (defined as `node-runtime/wasmer1_default`) and in `neard` (defined as `nearcore/wasmer1_default`).

**Review by olonho — APPROVED** [review]

**Pull request merged**
